packages/driver/src/cy/commands/querying/querying.ts

```diff
@@ -591,216 +229,172 @@ export default (Commands, Cypress, cy, state) => {
 
-      const setEl = ($el) => {
-        if (options.log === false) {
-          return
-        }
+      getOptions.withinSubject = subject[0] ?? subject
+      let $el = getFn()
 
-        consoleProps.Yielded = $dom.getElements($el)
-        consoleProps.Elements = $el != null ? $el.length : undefined
+      // .get() looks for elements *inside* the current subject, while contains() wants to also match the current
+      // subject itself if no child matches.
+      if (!$el.length) {
+        $el = (subject as JQuery).filter(selector)
+      }
 
-        options._log!.set({ $el })
+      if ($el.length) {
+        $el = $dom.getFirstDeepestElement($el)
       }
 
-      // find elements by the :cy-contains psuedo selector
-      // and any submit inputs with the attributeContainsWord selector
-      const selector = $dom.getContainsSelector(text, filter, options)
-
-      const resolveElements = () => {
-        const getOptions = _.extend({}, options, {
-          // error: getErr(text, phrase)
-          withinSubject: subject || state('withinSubject') || cy.$$('body'),
-          filter: true,
-          log: false,
-          // retry: false ## dont retry because we perform our own element validation
-          verify: false, // dont verify upcoming assertions, we do that ourselves
-        })
-
-        return cy.now('get', selector, getOptions).then(($el) => {
-          if ($el && $el.length) {
-            $el = $dom.getFirstDeepestElement($el)
+      log && log.set({
+        message: $utils.stringify(_.compact([filter, text])),
+        $el,
+        type: subject.is('body') ? 'parent' : 'child',
+        consoleProps: () => {
+          return {
+            Content: text,
+            'Applied To': $dom.getElements(subject),
+            Yielded: $el.get(0),
+            Elements: $el.length,
           }
+        },
+      })
 
-          setEl($el)
-
-          return cy.verifyUpcomingAssertions($el, options, {
-            onRetry: resolveElements,
-            onFail (err) {
-              switch (err.type) {
-                case 'length':
-                  if (err.expected > 1) {
-                    return $errUtils.throwErrByPath('contains.length_option', { onFail: options._log })
-                  }
-
-                  break
-                case 'existence':
-                  return err.message = getErr(err)
-                default:
-                  break
-              }
-
-              return null
-            },
-          })
-        })
-      }
-
-      return Promise
-      .try(resolveElements)
-    },
+      return $el
+    }
   })
 
-  Commands.add('shadow', { prevSubject: 'element' }, (subject, options) => {
-    const userOptions = options || {}
-
-    options = _.defaults({}, userOptions, { log: true })
-
-    const consoleProps: Record<string, any> = {
-      'Applied To': $dom.getElements(subject),
-    }
+  Commands._addQuery('shadow', function contains (userOptions: ShadowOptions = {}) {
```

Comment:
The way `_addQuery` is currently implemented, this changes .shadow from being a child command which requires an element subject to a dual command which an optional previous subject. This is new behavior that I don't anticipate is intentinal.

Reply:
Seems like this is handled below by `cy.ensureSubjectByType(subject, 'element')` instead of being part of the `_addQuery` functionality. Is that right?

Reply:
It defaults to dual/optional because I want to bypass the existing logic entirely. This goes pretty deep into the weeds of how `prevSubject` is implemented.

`prevSubject` is checked *before* invoking the command / query. But for queries, we instead want to throw an error *inside the retry loop* - the subject might (is expected to!) change between retries, and we want to run the validation each time.

So we use 'optional' to signal that Cypress shouldn't try running any assertions before invoking the command - any incoming subject is acceptable. The query will validate the incoming subject later, in the inner queryFn.

Reply:
Can you link through where this is? I have been pretty deep in this code recently, and my understanding of addCommand, is that we are actually [wrapping the commands callback ](https://github.com/cypress-io/cypress/blob/issue-7306-addQuery-custom-commands/packages/driver/src/cypress/cy.ts#L744)(which we aren't doing with _addQuery()) which enforces the previous subject check for us.

Reply:
Ok, went through the code and looks like my explanation is out of date - what I said is why I *originally* had optional / dual, but it no longer applies. Thank you for pressing on the matter!

`optional` and `dual` aren't needed at all, so I've removed them. Queries just plain old don't interact with the concept of `prevSubject` / `type` anymore - they follow their own codepath which bypasses them entirely, and my above explanation was badly out of date.